### DISCUSSION TITLE
Scope of project and configuration management?

### DISCUSSION THREAD

**Discussion by VorpalBlade** [steven-omaha/pacdef#51] [General]
## Background 

I have been using (mostly happily) [aconfmgr](https://github.com/CyberShadow/aconfmgr) to manage installed packages and setting files on my Arch Linux machines.

However I do have to use Debian/Ubuntu based distros at work. This naturally led me to consider how hard it would be to write it myself to add optional multi-distro support, if I would rewrite it in Rust, etc.

As part of searching I came across this crate, which seems to solve some of what I want (the package management bit). 

## Actual question

Is full on system package and config management something you are interested in/planning on doing for pacdef? Or is it out of scope? If it is out of scope, is it possible to structure things such that code reuse of a public API is possible?

As I see it, what I'm interested in can be split into multiple logical steps, some of which are missing:

* System discovery: based on host name, detected hardware etc decide which sets of features apply to the current system (this most likely requires a Turing complete config language).
* Install / uninstall packages based on those detected features (pacdef does this part already)
* Apply patches to stock configuration files (or sometimes overwrite them entirely) based on configuration files that can use the previous feature info. Definitely needs a Turing complete language to be able to handle things like managing /etc/shadow and /etc/passwd that will have different contents depending on what packages are installed (also I don't want to commit password hashes into my config repo, so I pull them from the system instead). Then there is also some config files (fstab for example) that are inherently host-specific and you basically need to manage on a per-host basis. 

One nice thing with aconfmgr is that it can also save the system to your config file (in that case it puts it in a 99-unsoeted.sh file). This is far better than something like ansible, which just manages the things you told it to, instead of trying to capture the entire system state. This is also something I would want.

(I'm under no illusion that I could use the same config file patches on Debian and Arch, but it be nice to have uniform tooling, and some things such as patching nanorc could be reused.)

**Comment by steven-omaha** [comment]
Thanks for your post!

> Is full on system package and config management something you are interested in/planning on doing for pacdef? Or is it out of scope?

It is unfortunately out of scope. I would like to follow the basic unix principle "do one thing and do it well".

> is it possible to structure things such that code reuse of a public API is possible?

That should in principle already be possible. This repo holds three crates. `pacdef` (binary application and thin wrapper), `pacdef_core` (actual library) and `pacdef_macros` (procedural macros for core). I'm exposing some, but not too many features in `pacdef_core`. If you would like to know details, you could start here https://docs.rs/pacdef_core/1.4.0/pacdef_core/ or follow the entrypoint from the main program into the library for an assumed action to get a feel how everything works. I can expose additional API if that helps the open source community. We could talk in more detail about this and figure it out together.

> System discovery: based on host name, detected hardware etc decide which sets of features apply to the current system (this most likely requires a Turing complete config language).

I have a personal dotfile manager that isn't public, where I deploy configs on the actual hostname. Intuitively, I would assume that once you look at what hardware is present, your project's complexity will go through the roof.

> Apply patches to stock configuration files (or sometimes overwrite them entirely) based on configuration files that can use the previous feature info

Here I started to think that what you actually want is the Nix package manager. It is in principle possible to deploy this to other distributions if your employer requires Debian. I have decided that this is not for me, but have you checked whether this could fit your needs?

**Comment by VorpalBlade** [comment]
> Intuitively, I would assume that once you look at what hardware is present, your project's complexity will go through the roof.

If I want to support this in general yes. I was rather thinking providing a list of found PCI devices etc to the user and let them script what is relevant. E.g. from my asysconf configuration I have:

```bash
	vga_controller="$(lspci | grep -E '3D |VGA ')"
	if grep -q NVIDIA <<< "$vga_controller"; then
		CUST_GPU_NVIDIA=1
	fi

	if grep -q Intel <<< "$vga_controller"; then
		CUST_GPU_INTEL=1
	fi

	if grep -q "Advanced Micro Devices" <<< "$vga_controller"; then
		CUST_GPU_AMD=1
	fi
```

So my plan is enabling the user to write the checks they want/need.

> Here I started to think that what you actually want is the Nix package manager.

Yes, kind of. I use and like Arch myself, and for work we use Ubuntu for dev environment (we build software for embedded, so that targets a yocto sysroot). Unfortunately when I looked at Nix there was a lot of complexity that didn't seem to be what I want.

I don't even *really* care about a declarative config, I'm perfectly happy having an imperative or rust-like language (rhai?) generate a set of operations to perform (packages that should be installed, config files that should differ from the standard ones, ...) and then be able to apply that. This is what aconfmgr I mentioned above does. It is written in bash (not the language I would choose myself for this, but it works just fine in practice).

If I make this project myself I imagine a rust core that deals with discovering current system state (installed packages, changed files) as well as applying the changes to match the wanted state. What changes are wanted is to be determined by an embedded scripting language of some sort (rhai? typescript? python?) that gets called with info on current system state. There is a bit of back and forth: ignored paths affects what gets scanned for example (so I will likely make it more structured than aconfmgr, which much parse the entire config first before scanning system state).

I initially envisioned having completely separate configs for arch and debian, but some things would be common (how to patch /etc/inputrc won't be distro specific for example), so maybe there is value in coming up with a decent way to have a shared section and distro specific sections. Food for thought.

EDIT: Oh and another reason for Debian support is Raspberry Pis. I have a few of them. Trying to remember to set up certain things the same of all of them is a pain.

**Reply by steven-omaha** [reply]
I see you have given this a good amount of thought. What I can offer you is to open up the API for the library when you need anything specific. Feel free to open an issue for that.

**Reply by VorpalBlade** [reply]
I do have a small followup question: What are the compatible aur helpers in the core library? I use [aurutils](https://github.com/aurutils/aurutils) myself, which works rather differently than most (creating a custom repo and having the built packages in there, it doesn't actually install any packages itself).

It really seems that your code is using the aur helper as a replacement binary for pacman, so I guess aurutils in particular won't work there.

Might be worth clarifying in the readme.

**Reply by steven-omaha** [reply]
See the "supported backends" section in the Readme. ;) supported right now are pacman wrappers like paru and yay.

**Reply by VorpalBlade** [reply]
I must have missed that. Sorry, my bad
